### DISCUSSION TITLE
Allow changing <Thickness> properties within .axaml

### DISCUSSION THREAD

**Discussion by Nickelony** [AvaloniaUI/Avalonia#12646] [Q&A] [answered]
I've been wondering, since I like to have shared Thicknesses all across my apps and I mainly use WPF these days, I have noticed that this code:
```xml
<StackPanel Spacing="{StaticResource MediumThicknessWidth}">
	<StackPanel.Margin>
		<Thickness
			Top="0"
			Bottom="0"
			Left="{StaticResource LargeThicknessWidth}"
			Right="{StaticResource LargeThicknessWidth}" />
	</StackPanel.Margin>
	...
</StackPanel>
```
where:
```xml
<sys:Double x:Key="MediumThicknessWidth">8</sys:Double>
<sys:Double x:Key="LargeThicknessWidth">16</sys:Double>
```
would work with WPF, but it doesn't work in Avalonia, since the `Top`, `Bottom`, `Left` and `Right` properties are inaccessible from the .axaml file.
Is there a workaround for this, or is it simply not possible to set each margin side separately using Resources?
If so, I think it would be great to expose those properties for more flexibility, while also matching WPF behavior.

**Comment by timunie** [comment]
DynamicResource would not work correctly in WPF wither as the properties are not observable, so after first init they will not report any changes. With DynamicResource this is unexpected, so Avalonia doesn't support it.

**Reply by Nickelony** [reply]
Yes, that's true, but what about using a {x:Static} in this case (which is what I usually use). I only need to set it once since changing margins in real time wouldn't make much sense. What I want is to keep all margins uniform across the app, and hardcoding them like "0,0,8,8" doesn't look right to me.

I have edited my example to be StaticResource to prevent further confusion.

Also what's the benefit of keeping the `Top`, `Bottom` etc. properties as `{ get; }` only, other than not allowing accidental `DynamicResource` binding? Maybe some people would like to use `DynamicResource` to find the target resource they need without having to include the containing `ResourceDictionary`? Even though it's lazy, it's really up to the programmer to prevent themselves from making such borderline irrelevant mistakes.

I think adding a setter to these properties would be very beneficial, especially since WPF already does that too.
**Example:** I have an entire WPF theme that I want to port to Avalonia, but I extensively use this type of code:
```xml
<!--  Thicknesses  -->
<Setter Property="BorderThickness" Value="{x:Static local:Defaults.BorderThickness}" />
<Setter Property="Padding">
	<Setter.Value>
		<Thickness
			Bottom="{x:Static local:Defaults.InputControl_DefaultPaddingWidth}"
			Left="{x:Static local:Defaults.InputControl_LargePaddingWidth}"
			Right="{x:Static local:Defaults.InputControl_LargePaddingWidth}"
			Top="{x:Static local:Defaults.InputControl_DefaultPaddingWidth}" />
	</Setter.Value>
</Setter>
```
Porting this to Avalonia is currently impossible and would require me to change almost all Padding / Margin / Thickness code.

**Reply by timunie** [reply]
I'd prefer this kind of approach: 

```xml
<App.Resources>
  <Thickness x:Key="Padding.LargeLeftLargeRight" />
</App.Resources>

<Setter Property="Padding" Value="{DynamicResource MyKeyToUse}" />
```

alternative:
Use a custom MarkupExtension to provide the padding
```xml
<Setter Property="Padding" Value="{my:ThicknessProvider Top={x:Static Defaults.Large}, Left=...}" />
```

**Reply by Nickelony** [reply, marked as answer]
Alright, I just wrote a `MarkupExtension` like this:
```cs
public class ThicknessProvider : MarkupExtension
{
	public double Left { get; set; }
	public double Top { get; set; }
	public double Right { get; set; }
	public double Bottom { get; set; }

	public override object ProvideValue(IServiceProvider serviceProvider)
		=> new Thickness(Left, Top, Right, Bottom);
}
```
which allows me to do this now:
```xml
<Separator>
	<Separator.Margin>
		<local:ThicknessProvider Left="{x:Static local:Defaults.LargeThicknessWidth}" Right="{x:Static local:Defaults.LargeThicknessWidth}" />
	</Separator.Margin>
</Separator>
```
alternatively:
```xml
<Separator Margin="{local:ThicknessProvider Left={x:Static local:Defaults.LargeThicknessWidth}, Right={x:Static local:Defaults.LargeThicknessWidth}}" />
```
This is actually perfect! Thank you! :)

**Reply by robloo** [reply]
I too missed this ability going from WPF to UWP (no other XAML framework since WPF has supported it).

Markup extension is a great idea @timunie! This should probably be converted to a Q&A and that above marked as the answer. It's definitely a workable solution with full functionality.

Alternatively, as mentioned, just declare the Thickness itself as a resource. That is limited but I've found that a lot of thicknesses are shared and you don't usually need full control of each component.

```xaml
<App.Resources>
  <Thickness x:Key="LargeLeftLargeRight">10,0,10,0</Thickness>
</App.Resources>

<Button Margin="{DynamicResource LargeLeftLargeRight}" />
```
